### ISSUE TITLE
Parse URLs in search for rel=me and RSS feeds

### ISSUE THREAD

**Issue by shiribailem** [friendica/friendica#13080]
### Describe the feature you'd like

The feature I'm imagining is being able to just drop a website URL into search and if it has a rel=me and/or an rss feed have them show up as results.

### Describe alternatives you've considered

Obviously you can manually search for them or see if their page tells you about it.

### Additional context

The basic gist is that this would do a simple parse of the page for basic patterns. It would probably do best as a core feature with an addon hook (so for instance it could readily be extended for non-standard cases, ie. an addon could hook in and automatically convert a reddit URL to it's rss feed or other similar cases).

* rel=me would pull the profile obviously
* links with type="application/rss+xml" or similar would list as RSS feeds
* look for hrefs ending in .rss
* look for <a> tags containing the word rss

This would allow most subscribable sites to just be added by homepage URL, it would also work as a sort of reverse of verification in that the rel=me could be used to subscribe based off of the homepage so you know you're subscribing to the right account.

**Comment by MrPetovan** [comment]
If you're looking to add a contact, the "Add New Contact" search field already looks for syndication feeds in the target page links. The other opportunities seems to fuzzy to me to be worthwhile to pursue since I expect them to yield many false positives.

**Comment by shiribailem** [comment]
I think the difference is that Add New Contact is looking for a single entry, right?

This suggestion is for search so it can potentially pull multiple entries. Example under the above guidelines for instance https://propublica.org would return their RSS feed (href including the word RSS) and their mastodon profile (rel=me).

It's also a sort of usability thing, allowing people to add these from search as opposed to going down into another section of the site. Basically for the typical "barely knows how to log into facebook" user that we want to draw in.

Also being a search the false positives wouldn't be that big a deal since it'd be a list of results. They could even be checked as part of the search.

I also wouldn't suggest this for anything other than manual searches.

(Though I'll note that I tried a wordpress site and propublica, both of which offer RSS feeds and neither was recognized, I'm guessing because the matching is probably only look for urls that end in .rss or some such)

**Comment by MrPetovan** [comment]
No, we're explicitly looking for `<link rel="alternate">` tags with the `type="application/rss+xml"` or `type="application/atom+xml"` attribute. On the other hand the ProPublica website exposes its feed URL like this:
```
<a class="btn btn-social social-main social-link-rss" href="https://www.propublica.org/feeds/propublica/main" id="rss-follow-pp-module">
	<svg class="icon" width="24" height="24"><use xlink:href="#icon-rss" /></svg>
	<span><abbr title="Rich Site Summary">RSS</abbr></span>
</a>
```
Which is a human-readable way of exposing it and we would have to make a specific rule to parse this.

**Comment by MrPetovan** [comment]
Wordpress blogs should work as the fee URLs are exposed in `<link>` tags by default.

**Comment by MrPetovan** [comment]
Regarding the search, I hear you. For contacts we are always searching for a unique result so it would require a significant rework of the search in general.

**Comment by shiribailem** [comment]
My sister's wordpress site (and storefront) doesn't recognize: https://tabletopteas.com/ and it is in a link tag.

And in the case of search, why would it be a significant rework? I imagine it would be a simple catch and reroute to another function. That is, like it does now it checks if it's an actual profile page first, but then if it fails that check it would reroute to the search call that outputs results based on the above steps. (Again to be clear, this would be from search, not add-contact)

**Comment by MrPetovan** [comment]
I'm able to pull the feed URL from entering the base URL in the "Add New Contact" field:
![image](https://user-images.githubusercontent.com/925415/236108303-302b0732-f4a9-408f-926b-176733f421e2.png)

**Comment by shiribailem** [comment]
Whelp, that turns that part into a bug/support request because on mine it just says it couldn't identify the network type...

**Comment by MrPetovan** [comment]
Just to confirm: you aren't entering the base URL in the search bar at the top of the screen, right?

And if you are using the "Add New Contact" field, you are entering the full URL including the scheme, right?

**Comment by shiribailem** [comment]
Yep, was just about to start doing some log digging to see if there's any errors.

![image](https://user-images.githubusercontent.com/5880760/236109313-031c1b0d-6a27-4654-aa44-9a020aadb7e0.png)

**Comment by MrPetovan** [comment]
You'd need to set up your Friendica log to the Info level during your attempts.

**Comment by shiribailem** [comment]
```
2023-05-04T04:08:52Z index [INFO]: Fetched for user 0 from https://tabletopteas.com/ returned 200 [] - {"file":"HTTPSignature.php","line":505,"function":"fetchRaw","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
2023-05-04T04:08:53Z index [NOTICE]: No valid network found {"url":"https://tabletopteas.com/","uid":0,"default":[],"update":true,"callstack":"Contact::getByURL (119), Follow::content (244), BaseModule::run (700), App::runFrontend (52)"} - {"file":"Contact.php","line":1340,"function":"getIdForURL","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
2023-05-04T04:08:53Z index [INFO]: Trying to fetch link {"uid":2,"uri":"https://tabletopteas.com/"} - {"file":"Item.php","line":3701,"function":"fetchByLink","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
2023-05-04T04:08:53Z index [INFO]: Fetched for user 2 from https://tabletopteas.com/ returned 200 [] - {"file":"HTTPSignature.php","line":505,"function":"fetchRaw","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
2023-05-04T04:08:53Z index [NOTICE]: Activity was not fetchable, aborting. {"url":"https://tabletopteas.com/","uid":2} - {"file":"Processor.php","line":1457,"function":"fetchCachedActivity","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
2023-05-04T04:08:53Z index [NOTICE]: Invalid url {"url":"https://tabletopteas.com/"} - {"file":"Diaspora.php","line":1157,"function":"fetchByURL","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
2023-05-04T04:08:53Z index [INFO]: Link not found {"uid":2,"uri":"https://tabletopteas.com/"} - {"file":"Item.php","line":3733,"function":"fetchByLink","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
2023-05-04T04:08:53Z index [INFO]: Trying to fetch link {"uid":0,"uri":"https://tabletopteas.com/"} - {"file":"Item.php","line":3701,"function":"fetchByLink","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
2023-05-04T04:08:54Z index [INFO]: Fetched for user 0 from https://tabletopteas.com/ returned 200 [] - {"file":"HTTPSignature.php","line":505,"function":"fetchRaw","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
2023-05-04T04:08:54Z index [NOTICE]: Activity was not fetchable, aborting. {"url":"https://tabletopteas.com/","uid":0} - {"file":"Processor.php","line":1457,"function":"fetchCachedActivity","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
2023-05-04T04:08:54Z index [NOTICE]: Invalid url {"url":"https://tabletopteas.com/"} - {"file":"Diaspora.php","line":1157,"function":"fetchByURL","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
2023-05-04T04:08:54Z index [INFO]: Link not found {"uid":0,"uri":"https://tabletopteas.com/"} - {"file":"Item.php","line":3733,"function":"fetchByLink","request-id":"64532fd31e883","uid":"c48362","process_id":1710871}
```

**Comment by MrPetovan** [comment]
Please provide the full log for `"request-id":"64532fd31e883"`.

**Comment by shiribailem** [comment]
that's everything in the log that matches that, is there some other log I could find that in?

**Comment by MrPetovan** [comment]
Something is missing, you should have either a `URI was not detectable` info log entry, or a `Probing start` info log entry before all the ones you shared already.

**Comment by shiribailem** [comment]
pulled it with debug and still no `URI was not detectable` or `Probing start`

**Comment by MrPetovan** [comment]
For reference, this is the log for a successful probe:
```
2023-05-04T05:10:06Z index [INFO]: Fetched for user 0 from https://tabletopteas.com/ returned 200 [] - {"file":"HTTPSignature.php","line":505,"function":"fetchRaw","request-id":"64533e2d72382","uid":"9027e7","process_id":17617}
2023-05-04T05:10:06Z index [INFO]: Probing start {"uri":"https://tabletopteas.com/"} - {"file":"Probe.php","line":723,"function":"detect","request-id":"64533e2d72382","uid":"9027e7","process_id":17617}
2023-05-04T05:10:07Z index [INFO]: No webfinger data retrievable {"url":"https://tabletopteas.com/.well-known/webfinger?resource=acct%3A%40tabletopteas.com"} - {"file":"Probe.php","line":976,"function":"webfinger","request-id":"64533e2d72382","uid":"9027e7","process_id":17617}
2023-05-04T05:10:07Z index [INFO]: No webfinger data retrievable {"url":"https://tabletopteas.com/.well-known/webfinger?resource=https%3A%2F%2Ftabletopteas.com%2F"} - {"file":"Probe.php","line":976,"function":"webfinger","request-id":"64533e2d72382","uid":"9027e7","process_id":17617}
2023-05-04T05:10:07Z index [INFO]: Probing {"host":"tabletopteas.com","ssl_url":"https://tabletopteas.com/.well-known/host-meta","url":"http://tabletopteas.com/.well-known/host-meta","callstack":"Probe::getWebfingerArray (730), Probe::detect (373), Probe::uri (1305), Contact::getIdForURL (317), Contact::getByURL (119), Follow::content (244), BaseModule::run (700), App::runFrontend (52)"} - {"file":"Probe.php","line":222,"function":"hostMeta","request-id":"64533e2d72382","uid":"9027e7","process_id":17617}
2023-05-04T05:10:08Z index [INFO]: No xrd object found {"host":"tabletopteas.com"} - {"file":"Probe.php","line":258,"function":"hostMeta","request-id":"64533e2d72382","uid":"9027e7","process_id":17617}
2023-05-04T05:10:09Z index [INFO]: Test Atom/RSS feed [] - {"file":"Feed.php","line":72,"function":"import","request-id":"64533e2d72382","uid":"9027e7","process_id":17617}
2023-05-04T05:10:09Z index [INFO]: There are no entries in this feed. [] - {"file":"Feed.php","line":275,"function":"import","request-id":"64533e2d72382","uid":"9027e7","process_id":17617}
2023-05-04T05:10:09Z index [INFO]: Test Atom/RSS feed [] - {"file":"Feed.php","line":72,"function":"import","request-id":"64533e2d72382","uid":"9027e7","process_id":17617}
2023-05-04T05:10:10Z index [INFO]: Contact inserted {"id":24231,"url":"https://tabletopteas.com/","uid":0} - {"file":"Contact.php","line":1386,"function":"getIdForURL","request-id":"64533e2d72382","uid":"9027e7","process_id":17617}
2023-05-04T05:10:11Z index [INFO]: Contact not found, update not desired {"url":"c","uid":0,"update":false} - {"file":"Contact.php","line":1298,"function":"getIdForURL","request-id":"64533e2d72382","uid":"9027e7","process_id":17617}
```

**Comment by shiribailem** [comment]
I am at a loss as to why it's not working... I've followed rss feeds before, but it looks like it's not even trying now?

**Comment by MrPetovan** [comment]
Would you be able to do some deep digging and add some bespoke logging calls to your server files?

If yes, please add the following statement on line 1304 of the `src/Model/Contact.php` file:
```php
Logger::debug('Probe::uri condition', ['default' => $default, 'update' => is_null($update) ? 'null' : ($update ? 'true' : 'false')]);
```

**Comment by shiribailem** [comment]
Here's the output of that line when I tried:

`2023-05-05T02:21:20Z index [DEBUG]: Probe::uri condition {"default":[],"update":"true"} - {"file":"Contact.php","line":1304,"function":"getIdForURL","request-id":"6454682046872","uid":"c759e0","process_id":1788948}`

**Comment by MrPetovan** [comment]
Thanks, this doesn't make any sense to me at all.

Please remove the previous statement and add the following on line 703 of the `src/Network/Probe.php` file:
```php
Logger::debug('Meddlesome addons', ['hookData' => $hookData]);
```

**Comment by shiribailem** [comment]
```
2023-05-05T02:58:43Z index [DEBUG]: Meddlesome addons {"hookData":{"uri":"https://tabletopteas.com/","network":"","uid":0,"result":[]}} - {"file":"Probe.php","line":703,"function":"detect","request-id":"645470e2c39d2","uid":"f50626","process_id":1789870}

2023-05-05T02:58:43Z index [NOTICE]: No valid network found {"url":"https://tabletopteas.com/","uid":0,"default":[],"update":true,"callstack":"Contact::getByURL (119), Follow::content (244), BaseModule::run (700), App::runFrontend (52)"} - {"file":"Contact.php","line":1341,"function":"getIdForURL","request-id":"645470e2c39d2","uid":"f50626","process_id":1789870}

2023-05-05T02:58:43Z index [DEBUG]: Meddlesome addons {"hookData":{"uri":"https://tabletopteas.com/","network":"mail","uid":2,"result":null}} - {"file":"Probe.php","line":703,"function":"detect","request-id":"645470e2c39d2","uid":"f50626","process_id":1789870}

2023-05-05T02:58:43Z index [INFO]: Trying to fetch link {"uid":2,"uri":"https://tabletopteas.com/"} - {"file":"Item.php","line":3701,"function":"fetchByLink","request-id":"645470e2c39d2","uid":"f50626","process_id":1789870}
```

**Comment by shiribailem** [comment]
I appreciate it's at least an *interesting* behavior lol.

I'll note that the only non-standard file edit I've made to the files is just patching the one line of mailstream so it stops crashing the addon setting page (even though it's not enabled at all). Other than that everything is either settings in Friendica or outside services (php-fpm, nginx)

**Comment by MrPetovan** [comment]
> ```
> 2023-05-05T02:58:43Z index [DEBUG]: Meddlesome addons {"hookData": {"uri":"https://tabletopteas.com/","network":"","uid":0,"result":[]}} - {"file":"Probe.php","line":703,"function":"detect","request-id":"645470e2c39d2","uid":"f50626","process_id":1789870}
> ```

BINGO!

Addons with probing capabilities currently should return `null` if they couldn't probe the provided URL, not empty array. I'll have a patch soon, thank you for your reactivity!

**Comment by shiribailem** [comment]
Is this a bug introduced in 2023.04-1? I was able to follow rss feeds before (and those are still updating)? (I don't know if this started with the change, but it's the only thing I can think to point to)

Even doing some programming myself I find it weird how much debugging a large project feels like magic because these bugs have to be inconsistent somehow... and now I'm having flashbacks to "Yes! Everything works perfectly, now time to show my boss!"

**Comment by MrPetovan** [comment]
It looks like the recent changes to the addons introduced by @annando stunted the contact addition routine. Addons registering the `probe_detect` hook are passed a data array they should set the `result` key of if and only if they have a successful probe result. However the tumblr and twitter addon have been setting this key with empty arrays instead, which still make Friendica believe they got a probe hit. This short-circuits the rest of the contact detection without returning a meaningful result.

I'm debating whether to relax the type expectation, maybe an empty array should also be considered as a non-result. The intent was that if the probe was targeting a specific network, an empty array would mean "probe was successful, no results were found". But evidently it didn't pan out so far.

On the other hand if I straighten the addons, they will be a good example to follow in further addons, reducing the likelihood of another fluke like this.

**Comment by shiribailem** [comment]
Given that admins might potentially install custom addons, I would personally opt for accept it but present a debug log entry for "out of spec".

Though for official addons this feels like something that could be incorporated into a unit test. But I imagine Friendica was never designed with unit testing in mind and I know how hard that is to add after the fact...

**Comment by MrPetovan** [comment]
So, the intent was:
- `null`: Tentative probe unsuccessful
- `[]`: Authoritative probe unsuccessful
- Filled array: Probe successful

However, so far the tumblr and twitter addons were returning `[]` for all the unsuccessful cases, which we can't guess is wrong from Friendica core.

> Though for official addons this feels like something that could be incorporated into a unit test. But I imagine Friendica was never designed with unit testing in mind and I know how hard that is to add after the fact...

🥲

**Comment by MrPetovan** [comment]
Unsuccessful RSS probe addressed in https://git.friendi.ca/friendica/friendica-addons/pulls/1382

**Comment by shiribailem** [comment]
Confirmed that it fixed that problem, though it now circles me back to the original feature request: expanding detection of RSS feeds and identifying rel=me for matching pages to profiles in a search/attempt-to-add.

Quite a few places don't really tag their rss feeds and just put a regular link on their site labeled RSS Feed. So I think it's not a big leap for the RSS probe to regex for RSS inside an 'a' tag.

**Comment by MrPetovan** [comment]
[Zalgo Says Don't Parse HTML With Regex](https://stackoverflow.com/a/1732454)

**Comment by shiribailem** [comment]
fair, but it's still a simple check, few a tags are going to have large quantities of text inside of them, and if they do it's reasonable to skip them (who the hell would link to an rss feed with a whole paragraph?)
